### PR TITLE
Avoid having more than one open submissions in a project

### PR DESCRIPTION
**Pull request by aaclan-ebi**

### CODE REVIEW

File: src/main/java/org/humancellatlas/ingest/core/exception/MultipleOpenSubmissions.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package org.humancellatlas.ingest.core.exception;
+
+public class MultipleOpenSubmissions extends RuntimeException {
```
Comment (prabh-t) [inline src/main/java/org/humancellatlas/ingest/core/exception/MultipleOpenSubmissions.java]: please append "Exception" to the class name to make it clear it is an exception class.

File: src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.stream.Collectors;
+
+
+@Component
+@RequiredArgsConstructor
+@Getter
+public class ProjectChangeListener extends AbstractMongoEventListener<Project> {
+    @Autowired
+    @NonNull
+    private final MessageRouter messageRouter;
```
Comment (prabh-t) [inline src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java]: remove unused dependency wired here

File: src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    protected Logger getLog() {
+        return log;
+    }
+
+    @Override
+    public void onBeforeSave(BeforeSaveEvent<Project> event) {
```
Comment (prabh-t) [inline src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java]: forget about my onBeforeCreate comment. the event doesn't seem to be in the latest spring-data mongodb lib anymore. Alternatively we could use ValidatingRepositoryEventListener for this type of check.

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java]: I originally used ValidatingRepositoryEventListener but from my testing, it doesn't seem to work for dbref links updates.

File: src/main/java/org/humancellatlas/ingest/core/exception/MultipleOpenSubmissions.java (hunk 1 above)
Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/core/exception/MultipleOpenSubmissions.java]: Renamed. Thanks!

File: src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java (hunk 2 above)
Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/project/ProjectChangeListener.java]: Good catch! Thanks!

File: src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java [hunk 4]
```diff
@@ -120,13 +109,18 @@ public SubmissionEnvelope createUpdateSubmissionEnvelope() {
...
         submissionEnvelopeCreateHandler.setUuid(updateSubmissionEnvelope);
         updateSubmissionEnvelope.setIsUpdate(true);
-        SubmissionEnvelope insertedUpdateSubmissionEnvelope = submissionEnvelopeRepository.insert(updateSubmissionEnvelope);
-        submissionEnvelopeCreateHandler.handleSubmissionEnvelopeCreation(updateSubmissionEnvelope);
+        SubmissionEnvelope insertedUpdateSubmissionEnvelope = createSubmissionEnvelope(updateSubmissionEnvelope);
+        return insertedUpdateSubmissionEnvelope;
+    }
+
+    public SubmissionEnvelope createSubmissionEnvelope(SubmissionEnvelope submissionEnvelope) {
+        SubmissionEnvelope insertedUpdateSubmissionEnvelope = submissionEnvelopeRepository.insert(submissionEnvelope);
```
Comment (rolando-ebi) [inline src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java]: `insertedUpdateSubmissionEnvelope` - does this piece of code specifically apply to update submissions? If so the method name should be updated

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java]: do you mean reverted?

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java]: oh i get what you mean.

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java]: i think it should be insertedSubmissionEnvelope

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelopeService.java]: done, thanks!

**Pull request merged**